contracts/DivviRegistry.sol

```diff
@@ -0,0 +1,189 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.24;
+
+import {AccessControlDefaultAdminRulesUpgradeable} from '@openzeppelin/contracts-upgradeable/access/extensions/AccessControlDefaultAdminRulesUpgradeable.sol';
+import {Initializable} from '@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol';
+import {UUPSUpgradeable} from '@openzeppelin/contracts-upgradeable/proxy/utils/UUPSUpgradeable.sol';
+
+/**
+ * @title DivviRegistry
+ * @notice A registry contract for managing Divvi entities and agreements
+ */
+contract DivviRegistry is
+  Initializable,
+  AccessControlDefaultAdminRulesUpgradeable,
+  UUPSUpgradeable
+{
+  // Entities storage
+  mapping(address => bool) private _entities; // entity => true (if entity exists)
+
+  // Agreement storage
+  mapping(bytes32 => bool) private _agreements; // keccak256(provider, consumer) => true (if agreement exists)
+  mapping(address => bool) private _requiresApproval; // entity => boolean (if entity requires approval)
```

Comment:
How about grouping `exists` and `requiresApproval` into an `EntityData` struct?

This could make the entity data a bit more organized and potentially saves some gas due to storage packing when accessing both flags.

```suggestion
  struct EntityData {
      bool exists;
      bool requiresApproval;
      // fields can be added here in a future upgrade if needed
      // this is upgrade safe as long as `EntityData` is only used in a mapping
}

  // Entities storage
  mapping(address => EntityData) private _entities;
  // Agreement storage
  mapping(bytes32 => bool) private _agreements; // keccak256(provider, consumer) => true (if agreement exists)
```

Reply:
yes that seems more neat, updated!